### PR TITLE
ACCUMULO-4647 - start of contributor guide

### PR DESCRIPTION
**Pull request by dlmarion**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -832,6 +832,7 @@
               <exclude>nbproject/**</exclude>
               <exclude>nb-configuration.xml</exclude>
               <exclude>nbactions.xml</exclude>
+              <exclude>CONTRIBUTING.md</exclude>
```
Comment (mikewalch) [inline pom.xml]: You can add the license header using markdown comments (i.e  `<!-- comment -->`) and not have exclude this file from RAT.  See the `README.md`

**Comment by joshelser** [comment]
@dlmarion did you look to https://accumulo.apache.org/contributor/source at all (I think this was mentioned in the thread yesterday)? At a glance, most of the info on that existing page wouldn't apply, but there might be some nice things to include (e.g. public API, checkstyle/formatter).

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,25 @@
+# Contributors Guide
+
+ If you believe that you have found a bug, please search for an existing [issue](https://issues.apache.org/jira/browse/accumulo) to see if it has already been reported. If you would like to add a new feature to Accumulo, please send an email with your idea to the [user](mailto:user@accumulo.apache.org) mail list. If it's appropriate, then we will create a ticket and assign it to you.
+
+## Development
+
+- Ensure that your work targets the correct branch
+- Add / update unit and integration tests
+- Ensure that Accumulo builds cleanly using the command: `mvn clean verify`
```
Comment (joshelser) [inline CONTRIBUTING.md]: I think the intent is that `mvn verify` passes before the change is submitted, but this reads as those devs should always use that (when in practice, `mvn package` might be just fine).

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+## Patch Submission
+
+- Before submission please squash your commits using a message that contains the JIRA issue number and a description of the changes.
```
Comment (joshelser) [inline CONTRIBUTING.md]: How about: "using a message that starts with the JIRA issue number". I could see the current text being misconstrued (e.g. `foo bar baz ACCUMULO-XXXX` or `XXXX foo bar baz`)

We're quite consistent with commit messages being `ACCUMULO-XXXX foo bar baz` (with more text after a line break)

**Comment by dlmarion** [comment]
Made modifications based on comments from @mikewalch  and @joshelser.

**Comment by mikewalch** [comment]
+1 Updates look good to me.

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+## Development
+
+- See the [Developer's Guide](https://accumulo.apache.org/contributor/source) for information regarding common build commands, IDE setup and more.
+- Ensure that your work targets the correct branch
+- Add / update unit and integration tests
+
+## Patch Submission
+
+- Ensure that Accumulo builds cleanly before submiting your patch using the command: `mvn clean verify`
```
Comment (keith-turner) [inline CONTRIBUTING.md]: Do we really want to recommend/require `mvn clean verify`?  That takes a long time.  What about `mvn clean verify -Psunny` of `mvn clean verify -DskipITs`?

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+ If you believe that you have found a bug, please search for an existing [issue](https://issues.apache.org/jira/browse/accumulo) to see if it has already been reported. If you would like to add a new feature to Accumulo, please send an email with your idea to the [dev](mailto:dev@accumulo.apache.org) mail list. If it's appropriate, then we will create a ticket and assign it to you.
+
+## Development
+
+- See the [Developer's Guide](https://accumulo.apache.org/contributor/source) for information regarding common build commands, IDE setup and more.
+- Ensure that your work targets the correct branch
+- Add / update unit and integration tests
+
+## Patch Submission
```
Comment (keith-turner) [inline CONTRIBUTING.md]: Unfortunately, the one that the automated code formatting does not over is imports.  For the Fluo, the docs explicitly state how imports should be organized (and how to do that in different IDEs).  Should that be added here?

File: CONTRIBUTING.md (hunk 4 above)
Comment (dlmarion) [inline CONTRIBUTING.md]: Do we want to skip ITs if they create or update one? I think we would want the command that runs the fastest that also completes all the checks. Not sure what that is...

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+- Ensure that Accumulo builds cleanly before submiting your patch using the command: `mvn clean verify`
+- Before submission please squash your commits using a message that starts with the JIRA issue number and a description of the changes.
+- Patches should be submitted in the form of Pull Requests to the Apache Accumulo GitHub [repository](https://github.com/apache/accumulo/) or to the [Review Board](https://reviews.apache.org) accumulo repository.
+
+## Review
+
+- While everyone is welcome to review patches, at least one committer needs to review and approve your patch.
```
Comment (mjwall) [inline CONTRIBUTING.md]: Are we moving from Commit then Review to Review then Commit with this line?  Or is it clear that this only applied to contributors and potential contributors.  Perhaps that warrants another DISCUSS thread.

File: CONTRIBUTING.md (hunk 4 above)
Comment (keith-turner) [inline CONTRIBUTING.md]: Personally, I think we should require/suggest a quick command before submission.  Let Jenkins and Travis run longer commands and report to the PR.  I think requiring running a 2 to 3 hour command discourages PRs.  I suggest `mvn clean verify -DskipITs` as a good minimum for PRs.

File: CONTRIBUTING.md (hunk 5 above)
Comment (dlmarion) [inline CONTRIBUTING.md]: I thought it was possible for the code formatting to organize the imports.

File: CONTRIBUTING.md (hunk 6 above)
Comment (mjwall) [inline CONTRIBUTING.md]: "Perhaps that warrants another DISCUSS thread."  - what I mean is that if this becomes a discussion on RTC vs CTR, I don't want do that on github as part of this PR

Comment (keith-turner) [inline CONTRIBUTING.md]: Personally I feel that `While everyone is welcome to review patches` has a slightly negative connotation.  Could change to the following.

```
We welcome reviews from anyone.  Any commiter can approve and merge a PR.
```

Comment (keith-turner) [inline CONTRIBUTING.md]: Oh, the last sentence I suggested I think makes things more clearly `Any commiter can approve and merge a PR.`.

I think we should write this guide such that it clearly implies CTR, as that is the current agreed upon way of working.

File: CONTRIBUTING.md (hunk 4 above)
Comment (dlmarion) [inline CONTRIBUTING.md]: ok, I'll make that change.

File: CONTRIBUTING.md (hunk 6 above)
Comment (dlmarion) [inline CONTRIBUTING.md]: I can make that change.

File: CONTRIBUTING.md (hunk 4 above)
Comment (keith-turner) [inline CONTRIBUTING.md]: Thinking about this a little more, there is the mvn command we would appreciate people running before submitting a PR.  Then there mvn command we would like committers to run before git push.  These don't have to be the same command.

**Comment by dlmarion** [comment]
Merged manually

**Pull request closed without merging**